server/app/jobs/certificate_renew_job.rb

```diff
@@ -0,0 +1,91 @@
+class CertificateRenewJob
+  include Celluloid
+  include Logging
+  include CurrentLeader
+  include WaitHelper
+
+  RENEW_INTERVAL = 5 * 60 # Run the renew check every 5 mins
+
+  def initialize(perform = true)
+    async.perform if perform
+  end
+
+  def perform
+    # TODO Maybe sleep a while first to allow everything else to start up before running this loop?
+    info 'starting to watch certificate renewals'
+    every(RENEW_INTERVAL) do
+      if leader?
+        renew_certificates
+      end
+    end
+  end
+
+  def renew_certificates
+    Grid.each do |grid|
+      grid.certificates.each do |cert|
```

Comment:
`Certificate.where(valid_until < Time.now + RENEW_TRESHOLD)` - cheap optimization

The `can_renew?` probably still needs to be in the loop, because you can't query across relations... should maybe raise an error, although there's nothing useful the server can do if it's unable to renew an expiring cert.

Reply:
👍